### PR TITLE
DROID-2378 Membership | Enhancement | Introduce incentives on join-space screen 

### PR DESCRIPTION
**Pull request by konstantiniiv**
---

- [x] I understand that contributing to this repository will require me to agree with the [CLA](https://github.com/anyproto/.github/blob/main/docs/CLA.md)

---

### What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI

### Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

### Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 [tech-docs](https://github.com/anyproto/tech-docs)
- [ ] 🙅 no documentation needed

### CODE REVIEW

File: core-models/src/main/java/com/anytypeio/anytype/core_models/ext/MembershipExt.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import com.anytypeio.anytype.core_models.membership.MembershipUpgradeReason
+import com.anytypeio.anytype.core_models.membership.TierId
+
+fun TierId.isPossibleToUpgrade(
+    reason: MembershipUpgradeReason
+): Boolean {
+    return when (reason) {
+        MembershipUpgradeReason.NumberOfEditors -> isPossibleToUpgradeNumberOfSpaceMembers()
+        MembershipUpgradeReason.NumberOfReaders -> isPossibleToUpgradeNumberOfSpaceMembers()
+        MembershipUpgradeReason.NumberOfSharedSpaces -> false
```
Comment (uburoiubu) [inline core-models/src/main/java/com/anytypeio/anytype/core_models/ext/MembershipExt.kt:13]: please add comment why it is false here.

File: core-ui/src/main/java/com/anytypeio/anytype/core_ui/features/multiplayer/Request.kt [hunk 2]
```diff
@@ -20,141 +25,242 @@ import androidx.compose.ui.res.stringResource
...
 @Composable
-private fun CommentView() {
-    Box(
-        modifier = Modifier
-            .height(IntrinsicSize.Min)
-            .padding(start = 20.dp, end = 20.dp)
-            .background(
-                color = colorResource(id = R.color.shape_tertiary),
-                shape = RoundedCornerShape(4.dp)
+private fun Buttons(newMember: Id,
```
Comment (uburoiubu) [inline core-ui/src/main/java/com/anytypeio/anytype/core_ui/features/multiplayer/Request.kt:174]: please move newMember to the next line

File: presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/ShareSpaceViewModel.kt [hunk 3]
```diff
@@ -127,7 +127,8 @@ class ShareSpaceViewModel(
                 )
             }.collect { (spaceResponse, membersResponse, isCurrentUserOwner) ->
                 val spaceView = spaceResponse.toSpaceView()
-                val spaceMembers = membersResponse.toSpaceMembers()
+                val spaceMembers
+                = membersResponse.toSpaceMembers().sortedByDescending { it.status == ParticipantStatus.JOINING }
```
Comment (uburoiubu) [inline presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/ShareSpaceViewModel.kt:131]: let's also add `ParticipantStatus.Removing`

File: presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt [hunk 4]
```diff
@@ -43,108 +47,60 @@ class SpaceJoinRequestViewModel(
...
+    private val _spaceMembers = MutableStateFlow<List<ObjectWrapper.SpaceMember>>(mutableListOf())
+    private val _newMember = MutableStateFlow<ObjectWrapper.SpaceMember?>(null)
+    private val _viewState = MutableStateFlow<ViewState>(ViewState.Init)
+    private val _activeTier = MutableStateFlow<TierId?>(null)
+    val viewState: StateFlow<ViewState> = _viewState
 
-    private val state = MutableStateFlow<State>(State.Init)
-
-    val viewState = MutableStateFlow<ViewState>(ViewState.Init)
+    private val _commands = MutableStateFlow<Command?>(null)
```
Comment (uburoiubu) [inline presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt:65]: consider using `MutableSharedFlow`

File: presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt [hunk 5]
```diff
@@ -155,147 +111,223 @@ class SpaceJoinRequestViewModel(
...
+    private fun createOtherState(
+        spaceView: ObjectWrapper.SpaceView,
+        spaceMembers: List<ObjectWrapper.SpaceMember>,
+        newMember: ObjectWrapper.SpaceMember?,
+        isCurrentUserOwner: Boolean
+    ): ViewState {
+        val canAddReaders = spaceView.canAddReaders(isCurrentUserOwner, spaceMembers)
+        val canAddWriters = spaceView.canAddWriters(isCurrentUserOwner, spaceMembers)
+        return when {
+            newMember == null -> ViewState.Error.EmptyMember
```
Comment (uburoiubu) [inline presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt:186]: please add error handling or remove this

File: presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt [hunk 6]
```diff
@@ -43,108 +47,60 @@ class SpaceJoinRequestViewModel(
...
     private val analytics: Analytics,
-    private val userPermissionProvider: UserPermissionProvider
+    private val userPermissionProvider: UserPermissionProvider,
+    private val spaceViewSubscriptionContainer: SpaceViewSubscriptionContainer,
+    private val membershipProvider: MembershipProvider
 ): BaseViewModel() {
 
     val isDismissed = MutableStateFlow(false)
     private val _isCurrentUserOwner = MutableStateFlow(false)
+    private val _spaceMembers = MutableStateFlow<List<ObjectWrapper.SpaceMember>>(mutableListOf())
```
Comment (uburoiubu) [inline presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt:59]: consider wrapping it into `Resultat<T>` class

File: presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt [hunk 7]
```diff
@@ -43,108 +47,60 @@ class SpaceJoinRequestViewModel(
...
+    private val _newMember = MutableStateFlow<ObjectWrapper.SpaceMember?>(null)
```
Comment (uburoiubu) [inline presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt:60]: idem

File: presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt [hunk 8]
```diff
@@ -43,108 +47,60 @@ class SpaceJoinRequestViewModel(
...
+    private val _viewState = MutableStateFlow<ViewState>(ViewState.Init)
```
Comment (uburoiubu) [inline presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt:61]: idem

File: presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt [hunk 9]
```diff
@@ -155,147 +111,223 @@ class SpaceJoinRequestViewModel(
...
+        newMember: ObjectWrapper.SpaceMember?,
+        isCurrentUserOwner: Boolean
+    ): ViewState {
+        val canAddReaders = spaceView.canAddReaders(isCurrentUserOwner, spaceMembers)
+        val canAddWriters = spaceView.canAddWriters(isCurrentUserOwner, spaceMembers)
+        return when {
+            newMember == null -> ViewState.Error.EmptyMember
+            !canAddReaders && !canAddWriters -> createViewStateWithButtons(spaceView, newMember, listOf(InviteButton.ADD_MORE_VIEWERS, InviteButton.ADD_MORE_EDITORS, InviteButton.REJECT))
+            else -> createViewStateWithButtons(
+                spaceView,
```
Comment (uburoiubu) [inline presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt:166]: please used named params

File: presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt [hunk 10]
```diff
@@ -155,147 +111,223 @@ class SpaceJoinRequestViewModel(
...
+                newMember,
```
Comment (uburoiubu) [inline presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt:167]: idem

File: presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt [hunk 11]
```diff
@@ -155,147 +111,223 @@ class SpaceJoinRequestViewModel(
...
-                            analytics.sendEvent(eventName = rejectInviteRequest)
-                            isDismissed.value = true
-                        },
-                        onFailure = { e ->
-                            Timber.e(e, "Error while approving join-space request").also {
-                                sendToast(e.msg())
-                            }
-                        }
-                    )
+            val filters = ObjectSearchConstants.filterNewMember(params.member)
```
Comment (uburoiubu) [inline presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt:248]: `vmParams`

File: presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt [hunk 12]
```diff
@@ -305,45 +337,62 @@ class SpaceJoinRequestViewModel(
...
-            val participants: List<ObjectWrapper.SpaceMember>
-        ) : State()
-        object Error : State()
-    }
+    data class Resultat(
+        val tierId: TierId,
+        val spaceView: ObjectWrapper.SpaceView,
+        val isCurrentUserOwner: Boolean,
+        val spaceMembers: List<ObjectWrapper.SpaceMember>,
+        val newMember: ObjectWrapper.SpaceMember?
```
Comment (uburoiubu) [inline presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt:367]: let's avoid NULLABLE!

File: app/src/main/java/com/anytypeio/anytype/ui/multiplayer/SpaceJoinRequestFragment.kt [hunk 13]
```diff
@@ -66,11 +59,27 @@ class SpaceJoinRequestFragment : BaseBottomSheetComposeFragment() {
...
+                        vm.commands.collect { command ->
+                            proceedWithCommand(command)
+                        }
+                    }
                 }
             }
         }
     }
 
+    private fun proceedWithCommand(command: SpaceJoinRequestViewModel.Command?) {
```
[the remark this answers is not in the document]
Comment (konstantiniiv) [reply app/src/main/java/com/anytypeio/anytype/ui/multiplayer/SpaceJoinRequestFragment.kt:72]: fixed

File: core-ui/src/main/java/com/anytypeio/anytype/core_ui/features/multiplayer/Request.kt (hunk 2 above)
Comment (konstantiniiv) [reply core-ui/src/main/java/com/anytypeio/anytype/core_ui/features/multiplayer/Request.kt:174]: done

File: presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/ShareSpaceViewModel.kt (hunk 3 above)
Comment (konstantiniiv) [reply presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/ShareSpaceViewModel.kt:131]: done

File: presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt (hunk 4 above)
Comment (konstantiniiv) [reply presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt:65]: done

File: presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt (hunk 5 above)
Comment (konstantiniiv) [reply presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt:186]: done

File: presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt (hunk 9 above)
Comment (konstantiniiv) [reply presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt:166]: done

File: presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt (hunk 10 above)
Comment (konstantiniiv) [reply presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt:167]: done

File: presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt [hunk 14]
```diff
@@ -155,147 +111,223 @@ class SpaceJoinRequestViewModel(
...
+    ): ViewState {
+        val canAddReaders = spaceView.canAddReaders(isCurrentUserOwner, spaceMembers)
+        val canAddWriters = spaceView.canAddWriters(isCurrentUserOwner, spaceMembers)
+        return when {
+            newMember == null -> ViewState.Error.EmptyMember
+            !canAddReaders && !canAddWriters -> createViewStateWithButtons(spaceView, newMember, listOf(InviteButton.ADD_MORE_VIEWERS, InviteButton.ADD_MORE_EDITORS, InviteButton.REJECT))
+            else -> createViewStateWithButtons(
+                spaceView,
+                newMember,
+                buildList {
```
[the remark this answers is not in the document]
Comment (konstantiniiv) [reply presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt:191]: done

File: presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt [hunk 15]
```diff
@@ -155,147 +111,223 @@ class SpaceJoinRequestViewModel(
...
+        spaceView: ObjectWrapper.SpaceView,
+        newMember: ObjectWrapper.SpaceMember,
+        buttons: List<InviteButton>
+    ): ViewState.Success {
+        return ViewState.Success(
+            spaceName = spaceView.name.orEmpty(),
+            newMember = newMember.identity,
+            newMemberName = newMember.name.orEmpty(),
+            icon = SpaceMemberIconView.icon(
+                obj = ObjectWrapper.SpaceMember(mapOf()),
```
[the remark this answers is not in the document]
Comment (konstantiniiv) [reply presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt:210]: done

File: core-models/src/main/java/com/anytypeio/anytype/core_models/ext/MembershipExt.kt (hunk 1 above)
Comment (konstantiniiv) [reply core-models/src/main/java/com/anytypeio/anytype/core_models/ext/MembershipExt.kt:15]: done

**Pull request merged**